### PR TITLE
[improvement](filecache) split file cache into sharding directories

### PR DESCRIPTION
**Pull request by AshinGau**
# Proposed changes

Save cached file segment into path like `cache_path / hash(filepath).substr(0, 3) / hash(filepath) / offset` to prevent too many directories in `cache_path`.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.09 seconds
 stream load tsv:          486 seconds loaded 74807831229 Bytes, about 146 MB/s
 stream load json:         35 seconds loaded 2358488459 Bytes, about 64 MB/s
 stream load orc:          69 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230215024529_clickbench_pr_96525.html

File: be/src/io/cache/block/block_lru_file_cache.cpp [hunk 1]
```diff
@@ -634,63 +637,113 @@ void LRUFileCache::remove(const Key& key, bool is_persistent, size_t offset,
...
+                            LOG(WARNING) << "Failed to create new version cached directory: "
+                                         << ec.message();
+                            continue;
+                        }
+                    }
+                    std::error_code ec;
+                    std::filesystem::rename(key_it->path(), key_prefix / cache_key, ec);
+                    if (ec) {
+                        LOG(WARNING)
+                                << "Failed to move old version cached directory: " << ec.message();
```
Comment (morningman) [inline be/src/io/cache/block/block_lru_file_cache.cpp:664]: If error, I think we should clean all cache to avoid dirty data?

Comment (AshinGau) [reply be/src/io/cache/block/block_lru_file_cache.cpp:664]: Dirty data has been removed in the following process `key_prefix_it->path().filename().native().size() != KEY_PREFIX_LENGTH`

File: be/src/io/cache/block/block_file_cache.cpp [hunk 2]
```diff
@@ -32,6 +32,9 @@ namespace fs = std::filesystem;
 namespace doris {
 namespace io {
 
+const std::string IFileCache::FILE_CACHE_VERSION = "2.0";
+const int IFileCache::KEY_PREFIX_LENGTH = 3;
```
Comment (platoneko) [inline be/src/io/cache/block/block_file_cache.cpp:36]: `const` -> `constexpr` is better

Comment (AshinGau) [reply be/src/io/cache/block/block_file_cache.cpp:36]: string can't use constexpr, I'll use constexpr for `KEY_PREFIX_LENGTH` in next PR.

**Review by morningman — APPROVED** [review]
LGTM

File: be/src/io/cache/block/block_file_cache.cpp (hunk 2 above)
Comment (platoneko) [reply be/src/io/cache/block/block_file_cache.cpp:36]: `std::string_view` can be constexpr[Lol]

**Pull request merged**